src/txrebroadcast.cpp

```diff
@@ -0,0 +1,141 @@
+
+    BlockAssembler::Options options;
+    options.nBlockMaxWeight = MAX_REBROADCAST_WEIGHT;
+    options.m_skip_inclusion_until = start_time - REBROADCAST_MIN_TX_AGE;
+    options.check_block_validity = false;
+    options.blockMinFeeRate = m_cached_fee_rate;
+
+    // Use CreateNewBlock to identify rebroadcast candidates
+    auto block_template = BlockAssembler(m_mempool, Params(), options)
+                          .CreateNewBlock(m_chainman.ActiveChainstate(), CScript());
+    std::chrono::microseconds after_CNB_time = GetTime<std::chrono::microseconds>();
+
+    for (const CTransactionRef& tx : block_template->block.vtx) {
+        if (tx->IsCoinBase()) continue;
+
+        uint256 txid = tx->GetHash();
+        uint256 wtxid = tx->GetWitnessHash();
+
+        // Check if we have previously rebroadcasted, decide if we will this
+        // round, and if so, record the attempt.
+        auto entry_it = m_attempt_tracker.find(wtxid);
+
+        if (entry_it == m_attempt_tracker.end()) {
+            // No existing entry, we will rebroadcast, so create a new one
+            RebroadcastEntry entry(start_time, wtxid);
+            m_attempt_tracker.insert(entry);
+        } else if (entry_it->m_count >= MAX_REBROADCAST_COUNT) {
+            // We have already rebroadcast this transaction the maximum number
+            // of times permitted, so skip rebroadcasting.
+            continue;
+        } else if (entry_it->m_last_attempt > start_time - MIN_REATTEMPT_INTERVAL) {
+            // We already rebroadcasted this in the past 4 hours. Even if we
+            // added it to the set, it would probably not get INVed to most
+            // peers due to filterInventoryKnown.
+            continue;
+        } else {
+            // We have rebroadcasted this transaction before, but will try
+            // again now.
+            RecordAttempt(entry_it);
+        }
+
+        // Add to set of rebroadcast candidates
+        rebroadcast_txs.push_back(TxIds(txid, wtxid));
+    }
+
+    TrimMaxRebroadcast();
+
+    std::chrono::microseconds delta1 = after_CNB_time - start_time;
+    std::chrono::microseconds delta2 = GetTime<std::chrono::microseconds>() - start_time;
+    LogPrint(BCLog::BENCH, "GetRebroadcastTransactions(): %d µs total, %d µs spent in CreateNewBlock.\n", delta1.count(), delta2.count());
+    LogPrint(BCLog::NET, "Queued %d transactions for attempted rebroadcast, filtered from %d candidates with cached fee rate of %s.\n", rebroadcast_txs.size(), block_template->block.vtx.size() - 1, m_cached_fee_rate.ToString(FeeEstimateMode::SAT_VB));
+
+    for (TxIds ids : rebroadcast_txs) {
+        LogPrint(BCLog::NET, "Attempting to rebroadcast txid: %s, wtxid: %s\n", ids.m_txid.ToString(), ids.m_wtxid.ToString());
+    }
+
+    return rebroadcast_txs;
+};
+
+void TxRebroadcastHandler::CacheMinRebroadcastFee()
+{
+    // Update stamp of chain tip on cache run
+    m_tip_at_cache_time = ::ChainActive().Tip();
```

Comment:
Avoid calls to the global `::ChainActive()`.

Reply:
done by making chainman a member